File: js/packages/mechanisms/casper/src/exact/server/scheme.ts
import {
  AssetAmount,
  Network,
  PaymentRequirements,
  Price,
  SchemeNetworkServer,
  MoneyParser,
  SupportedKind,
} from "@x402/core/types";
import { parseMoneyString } from "@x402/core/utils";
import { isValidAddress, isValidContractPackageHash, parseAmount } from "../../utils";

export const ErrNoDefaultAsset = "invalid_exact_casper_server_no_default_asset";
export const ErrInvalidAsset = "invalid_exact_casper_server_invalid_asset";
export const ErrInvalidPayTo = "invalid_exact_casper_server_invalid_payto";
export const ErrMissingTokenName = "invalid_exact_casper_server_missing_token_name";
export const ErrMissingTokenVersion = "invalid_exact_casper_server_missing_token_version";
export const ErrFailedToParseAmount = "invalid_exact_casper_server_failed_to_parse_amount";

/**
 * Build a lookup key for the asset decimals registry.
 *
 * @param network - The network identifier
 * @param asset - The asset contract package hash
 * @returns Registry key
 */
function assetDecimalsKey(network: Network, asset: string): string {
  return `${network}:${asset}`;
}

/**
 * Casper server implementation for the Exact payment scheme.
 */
export class ExactCasperScheme implements SchemeNetworkServer {
  readonly scheme = "exact";
  private moneyParsers: MoneyParser[] = [];
  private assetDecimals = new Map<string, number>();

  /**
   * Register a custom money parser in the parser chain.
   *
   * @param parser - Custom function to convert amount to AssetAmount (or null to skip)
   * @returns The server instance for chaining
   */
  registerMoneyParser(parser: MoneyParser): ExactCasperScheme {
    this.moneyParsers.push(parser);
    return this;
  }

  /**
   * Register the decimal precision for an asset on a network.
   *
   * @param network - The network identifier
   * @param asset - The asset contract package hash
   * @param decimals - Number of decimal places
   * @returns The server instance for chaining
   */
  registerAsset(network: Network, asset: string, decimals: number): ExactCasperScheme {
    this.assetDecimals.set(assetDecimalsKey(network, asset), decimals);
    return this;
  }

  /**
   * Get the decimal precision for an asset on a network.
   *
   * @param asset - The asset contract package hash
   * @param network - The network identifier
   * @returns Number of decimal places
   */
  getAssetDecimals(asset: string, network: Network): number {
    return this.assetDecimals.get(assetDecimalsKey(network, asset)) ?? 9;
  }

  /**
   * Parse a price into an asset amount.
   *
   * @param price - The price to parse
   * @param network - The network to use
   * @returns Promise resolving to the parsed asset amount
   */
  async parsePrice(price: Price, network: Network): Promise<AssetAmount> {
    if (typeof price === "object" && price !== null && "amount" in price) {
      if (!isValidContractPackageHash(price.asset)) {
        throw new Error(`${ErrInvalidAsset}: ${price.asset}`);
      }
      return {
        amount: price.amount,
        asset: price.asset,
        extra: price.extra || {},
      };
    }

    const amount = this.parseMoneyToDecimal(price);

    for (const parser of this.moneyParsers) {
      const result = await parser(amount, network);
      if (result !== null) {
        return result;
      }
    }

    throw new Error(`${ErrNoDefaultAsset}: no default asset configured for network ${network}`);
  }

  /**
   * Build payment requirements for this scheme/network combination.
   *
   * @param paymentRequirements - The base payment requirements
   * @param supportedKind - The supported kind from facilitator
   * @param extensionKeys - Extension keys supported by the facilitator
   * @returns Enhanced payment requirements
   */
  async enhancePaymentRequirements(
    paymentRequirements: PaymentRequirements,
    supportedKind: SupportedKind,
    extensionKeys: string[],
  ): Promise<PaymentRequirements> {
    if (!isValidContractPackageHash(paymentRequirements.asset)) {
      throw new Error(`${ErrInvalidAsset}: ${paymentRequirements.asset}`);
    }

    if (!isValidAddress(paymentRequirements.payTo)) {
      throw new Error(`${ErrInvalidPayTo}: ${paymentRequirements.payTo}`);
    }

    let amount = paymentRequirements.amount;
    if (amount.includes(".")) {
      const decimals = this.getAssetDecimals(
        paymentRequirements.asset,
        paymentRequirements.network,
      );
      try {
        amount = parseAmount(amount, decimals).toString();
      } catch (error) {
        const message = error instanceof Error ? error.message : String(error);
        throw new Error(`${ErrFailedToParseAmount}: ${message}`);
      }
    }

    const extra = { ...paymentRequirements.extra };

    if (typeof extra.name !== "string" || extra.name === "") {
      throw new Error(ErrMissingTokenName);
    }
    if (typeof extra.version !== "string" || extra.version === "") {
      throw new Error(ErrMissingTokenVersion);
    }

    if (supportedKind.extra) {
      for (const key of extensionKeys) {
        if (Object.hasOwn(supportedKind.extra, key)) {
          extra[key] = supportedKind.extra[key];
        }
      }
    }

    return {
      ...paymentRequirements,
      amount,
      extra,
    };
  }

  /**
   * Parse Money (string | number) to a decimal number.
   *
   * @param money - The money value to parse
   * @returns Decimal number
   */
  private parseMoneyToDecimal(money: string | number): number {
    if (typeof money === "number") {
      return money;
    }

    return parseMoneyString(money);
  }
}

File: go/x402/mechanisms/casper/exact/server/scheme.go
package server

import (
	"context"
	"errors"
	"fmt"
	"strconv"
	"strings"

	"casper_x402_facilitator/x402/mechanisms/casper"

	x402 "github.com/x402-foundation/x402/go"
	"github.com/x402-foundation/x402/go/types"
)

type assetKey struct {
	network string
	asset   string
}

type ExactCasperScheme struct {
	moneyParsers  []x402.MoneyParser
	assetDecimals map[assetKey]int
}

func NewExactCasperScheme() *ExactCasperScheme {
	return &ExactCasperScheme{assetDecimals: make(map[assetKey]int)}
}

func (s *ExactCasperScheme) Scheme() string {
	return casper.SchemeExact
}

func (s *ExactCasperScheme) RegisterMoneyParser(parser x402.MoneyParser) *ExactCasperScheme {
	s.moneyParsers = append(s.moneyParsers, parser)
	return s
}

func (s *ExactCasperScheme) RegisterAsset(network, asset string, decimals int) *ExactCasperScheme {
	s.assetDecimals[assetKey{network: network, asset: asset}] = decimals
	return s
}

func (s *ExactCasperScheme) GetAssetDecimals(asset string, network x402.Network) int {
	if d, ok := s.assetDecimals[assetKey{network: string(network), asset: asset}]; ok {
		return d
	}

	return 9
}

func (s *ExactCasperScheme) ParsePrice(price x402.Price, network x402.Network) (x402.AssetAmount, error) {
	if priceMap, ok := price.(map[string]interface{}); ok {
		amountValue, hasAmount := priceMap["amount"]
		if hasAmount {
			amount, ok := amountValue.(string)
			if !ok {
				return x402.AssetAmount{}, errors.New(ErrAmountMustBeString)
			}

			assetValue, ok := priceMap["asset"]
			if !ok {
				return x402.AssetAmount{}, fmt.Errorf("%s: asset field required", ErrInvalidAsset)
			}
			asset, ok := assetValue.(string)
			if !ok || !casper.IsValidContractPackageHash(asset) {
				return x402.AssetAmount{}, fmt.Errorf("%s: %v", ErrInvalidAsset, assetValue)
			}

			extra := make(map[string]interface{})
			if extraValue, hasExtra := priceMap["extra"]; hasExtra {
				if extraMap, ok := extraValue.(map[string]interface{}); ok {
					extra = extraMap
				}
			}

			return x402.AssetAmount{Amount: amount, Asset: asset, Extra: extra}, nil
		}
	}

	decimalAmount, err := parseMoneyToDecimal(price)
	if err != nil {
		return x402.AssetAmount{}, err
	}

	for _, parser := range s.moneyParsers {
		assetAmount, err := parser(decimalAmount, network)
		if err != nil {
			continue
		}
		if assetAmount != nil {
			return *assetAmount, nil
		}
	}

	return x402.AssetAmount{}, fmt.Errorf(
		"%s: no default asset configured for network %s; provide an explicit AssetAmount with 'amount', 'asset', and 'extra' fields, or register a MoneyParser",
		ErrNoDefaultAsset,
		network,
	)
}

func parseMoneyToDecimal(price x402.Price) (float64, error) {
	switch v := price.(type) {
	case string:
		clean := strings.TrimSpace(strings.TrimPrefix(strings.TrimSpace(v), "$"))
		f, err := strconv.ParseFloat(clean, 64)
		if err != nil {
			return 0, fmt.Errorf("%s: %q", ErrInvalidPriceFormat, v)
		}
		return f, nil
	case float64:
		return v, nil
	case int:
		return float64(v), nil
	case int64:
		return float64(v), nil
	default:
		return 0, fmt.Errorf("%s: unsupported type %T", ErrInvalidPriceFormat, price)
	}
}

func (s *ExactCasperScheme) EnhancePaymentRequirements(
	_ context.Context,
	requirements types.PaymentRequirements,
	supportedKind types.SupportedKind,
	extensionKeys []string,
) (types.PaymentRequirements, error) {
	if !casper.IsValidContractPackageHash(requirements.Asset) {
		return requirements, fmt.Errorf("%s: %s", ErrInvalidAsset, requirements.Asset)
	}

	if !casper.IsValidAddress(requirements.PayTo) {
		return requirements, fmt.Errorf("%s: %s", ErrInvalidPayTo, requirements.PayTo)
	}

	if strings.Contains(requirements.Amount, ".") {
		decimals := s.GetAssetDecimals(requirements.Asset, x402.Network(requirements.Network))
		parsedAmount, err := casper.ParseAmount(requirements.Amount, decimals)
		if err != nil {
			return requirements, fmt.Errorf("%s: %w", ErrFailedToParseAmount, err)
		}
		requirements.Amount = parsedAmount.String()
	}

	if requirements.Extra == nil {
		requirements.Extra = map[string]interface{}{}
	}

	if name, ok := requirements.Extra["name"].(string); !ok || name == "" {
		return requirements, errors.New(ErrMissingTokenName)
	}
	if version, ok := requirements.Extra["version"].(string); !ok || version == "" {
		return requirements, errors.New(ErrMissingTokenVersion)
	}

	if supportedKind.Extra != nil {
		for _, key := range extensionKeys {
			if value, ok := supportedKind.Extra[key]; ok {
				requirements.Extra[key] = value
			}
		}
	}

	return requirements, nil
}
